src/expr.cpp

```diff
@@ -6052,22 +6052,24 @@ static llvm::Value *lTypeConvAtomic(FunctionEmitContext *ctx, llvm::Value *exprV
         case AtomicType::TYPE_INT16:
         case AtomicType::TYPE_INT32:
         case AtomicType::TYPE_INT64:
-            cast = ctx->CastInst(llvm::Instruction::SIToFP, // signed int to float16
-                                 exprVal, targetType, cOpName);
+            cast = ctx->I2HCastInst(llvm::Instruction::SIToFP, exprVal, targetType, cOpName);
             break;
         case AtomicType::TYPE_UINT8:
         case AtomicType::TYPE_UINT16:
         case AtomicType::TYPE_UINT32:
         case AtomicType::TYPE_UINT64:
-            cast = ctx->CastInst(llvm::Instruction::UIToFP, // unsigned int to float16
-                                 exprVal, targetType, cOpName);
+            cast = ctx->I2HCastInst(llvm::Instruction::UIToFP, exprVal, targetType, cOpName);
             break;
         case AtomicType::TYPE_FLOAT16:
             // No-op cast.
             cast = exprVal;
             break;
         case AtomicType::TYPE_FLOAT:
-            cast = ctx->FPCastInst(exprVal, targetType, cOpName);
+            if (g->target->hasHalf()) {
+                cast = ctx->FPCastInst(exprVal, targetType, cOpName);
+            } else {
+                cast = ctx->F2HCastInst(exprVal, targetType, cOpName);
+            }
             break;
         case AtomicType::TYPE_DOUBLE:
             cast = ctx->FPCastInst(exprVal, targetType, cOpName);
```

Comment:
How the fpcast work when we cast from fp16 to double? Don't we need to use H2F cast before?

Reply:
So it means there no tests for float16 <-> double conversion. 

Yes we need. Fixed that.

Reply:
Please add at least one runtime test checking float16<->double conversion.

Reply:
I have added. It looks like it reveals the problem that compiler-rt mey generate calls to trunc or extend functions even for architecture with native half support, e.g., avx2-*. Probably, it is because it doesn't have double to float16 HW instruction.